File: Reqnroll/Chapter16/16.1-ScenarioResourcePattern/After/WIMP.Specs/StepDefinitions/AuthenticationStepDefinitions.cs
using Reqnroll;

using WIMP.App.RestApi;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class AuthenticationStepDefinitions(AuthenticationApiDriver authApiDriver)
{
    private TestActionResult<LoginResponse> loginResult = TestActionResult<LoginResponse>.NotExecuted;

    [When("the customer attempts to log in with a wrong password")]
    public async Task WhenTheCustomerAttemptsToLogInWithAWrongPassword()
    {
        loginResult = await authApiDriver.Login(DomainDefaults.CustomerName, DomainDefaults.WrongPassword).AttemptExecute();
    }

    [Then("the login should fail with {string}")]
    public void ThenTheLoginShouldFailWith(string expectedMessage)
    {
        loginResult.AssertFailedWithErrorMessageContains(expectedMessage);
    }
}

File: CucumberJVM/Chapter20/20.3-ParameterizedExecutionPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/AuthenticationStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.restapi.LoginResponse;
import com.wimp.app.specs.drivers.AuthenticationApiDriver;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.TestActionResult;
import io.cucumber.java.en.*;

public class AuthenticationStepDefinitions {
    private final AuthenticationApiDriver authenticationApiDriver;
    private TestActionResult<LoginResponse> loginResult;

    public AuthenticationStepDefinitions(AuthenticationApiDriver authenticationApiDriver) {
        this.authenticationApiDriver = authenticationApiDriver;
    }

    @When("the customer attempts to log in with a wrong password")
    public void theCustomerAttemptsToLogInWithAWrongPassword() throws Exception {
        loginResult = authenticationApiDriver.login(DomainDefaults.CUSTOMER_NAME, DomainDefaults.WRONG_PASSWORD).attemptExecute();
    }

    @Then("the login should fail with {string}")
    public void theLoginShouldFailWith(String expectedMessage) {
        loginResult.assertFailedWithErrorMessageContains(expectedMessage);
    }
}
